### PR TITLE
[Login] Allow connecting Jetpack when signing in using WPCom credentials

### PR DESCRIPTION
**Pull request by hichamboushaba**
Part of: #7307 

### Description
This PR implements iteration 2 of the Account Mismatch improvements project, this iteration consists of handling the connection to Jetpack to self-hosted sites when the initial login is using WPCom credentials, and the flow is as follows:

1. When the WPCom account is not connected to the current site, we check if it's self-hosted.
2. If it's self-hosted, then we allow connecting Jetpack, and we ask the user about their wpadmin credentials:
    a. if the wpadmin account is not connected to any WPCom account, then we open a WebView to handle the connection.
    b. if the wpadmin account shows that it's already connected to a different WPCom account, then we switch to login using this account, and we ask the user about its password.

Next Steps:
- Improve the WebView experience by auto signing to the WPCom account (basically replicate the function of the WPComWebViewFragment)

##### Preparations and common steps
1. Prepare a site that has two wpadmin accounts (let's call them account 1 and account 2)
2. Prepare two WPCom accounts (same let's call them account a and account b)
3. Connect wpadmin account 1 to Jetpack using WPCom account a.
5. Open the app, then sign in using the site address.
6. Continue the login using WPCom credentials, using the WPCom account b.
7. Notice the "Connect Jetpack to your account" button in the Account Mismatch screen.
8. Click the button, then continue according to the scenarios below.

##### Scenario 1: wpadmin account not connected
1. Sign in using the wpdamin account 2.
2. Confirm the app loads the Jetpack connection URL successfully and opens a WebView to continue login.
3. Continue the sign-in in the WebView.
4. Confirm the app logs you to the site successfully.

##### Scenario 2: wpadmin account not connected
1. Sign in using wpadmin account 1.
2. Confirm the app shows the email of WPCom account a, and asks for confirmation.
3. Continue the login using magic email or password.
4. Confirm the app logs you to the site successfully.

##### wpadmin account not connected to Jetpack:

https://user-images.githubusercontent.com/1657201/193765175-9b94ada2-3cd2-4d4c-8b52-5c4a6515e864.mov

##### wpadmin account connected to different account:

https://user-images.githubusercontent.com/1657201/193765499-ff1147b8-921b-49d4-bc73-ee199ee9271a.mov

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt [hunk 1]
```diff
@@ -353,19 +353,13 @@ class MainActivity :
     override fun onBackPressed() {
         AnalyticsTracker.trackBackPressed(this)
 
-        if (!isAtNavigationRoot()) {
-            // go no further if active fragment doesn't allow back press - we use this so fragments can
-            // provide confirmation before discarding the current action, such as adding an order note
-            getActiveChildFragment()?.let { fragment ->
-                if (fragment is BackPressListener && !(fragment as BackPressListener).onRequestAllowBackPress()) {
-                    return
-                }
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainActivity.kt:366]: This logic is a leftover from the old navigation workaround, and it's preventing the usage of the "onBackPressedDispatcher" API.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModel.kt [hunk 2]
```diff
@@ -266,22 +267,38 @@ class SitePickerViewModel @Inject constructor(
...
-            )
-        )
-        trackLoginEvent(currentStep = UnifiedLoginTracker.Step.WRONG_WP_ACCOUNT)
+    private fun showAccountMismatchScreen(url: String) = launch {
         if (event.value !is NavigateToAccountMismatchScreen) {
-            triggerEvent(
-                NavigateToAccountMismatchScreen(
-                    hasConnectedStores = sitePickerViewState.hasConnectedStores ?: false,
-                    siteUrl = url
+            sitePickerViewState = sitePickerViewState.copy(isSkeletonViewVisible = true)
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModel.kt]: NP. We could add `isPrimaryBtnVisible = false` to hide the continue button while the SitePicker is still loading.

**Comment by JorgeMucientes** [comment]
Nicely done @hichamboushaba. And thanks for adding the password field Composable. Also, I think some of the functions of `AccountMismatchRepository` are going to be useful for the re-work to be done for the login simplification project. We might end up having a more general purpose. 

I tested both scenarios, and I noticed an unexpected behaviour for scenario 1. At the end of the flow you end up back in the Account mismatch screen, even tough you are correctly logged in, so if you press back button you land on MyStore tab. See the screen recording: 

https://user-images.githubusercontent.com/2663464/194107805-c2f8fb84-f89c-4ae7-9972-7d872aa59fa9.mp4

**Comment by hichamboushaba** [comment]
> I noticed an unexpected behaviour for scenario 1.

Thanks @JorgeMucientes for the test, I confirmed the issue after running multiple tests, it doesn't happen always. The cause is that the `if` check I have to avoid triggering the navigation twice didn't work sometimes because the /site-info call took longer, which meant we submit two `NavigateToAccountMismatchScreen` events, so after coming back, we consume the second event and we show the screen a second time.
I pushed [5831b6a](https://github.com/woocommerce/woocommerce-android/pull/7471/commits/5831b6a5cc29e838921a3d742c4ed84bbc8b63b5) to fix this, it just moves the if check after the /site-info call.
I think we can improve things further by fixing the root cause: avoiding the multiple calls to `showAccountMismatchScreen`, but this would require more work to get it right.

Ready for another round.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModel.kt (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/sitepicker/SitePickerViewModel.kt]: Good point, done in [bebf204](https://github.com/woocommerce/woocommerce-android/pull/7471/commits/bebf204739c43dadd21f59b1e321a96b7d4eb194)

**Comment by JorgeMucientes** [comment]
Thanks for fixing the race condition @hichamboushaba. I've re tested it and works smoothly 🙌🏼 
Ready to merge :shipit:

**Review by JorgeMucientes — APPROVED** [review]

**Pull request merged**